include/aspect/simulator.h

```diff
@@ -546,9 +567,74 @@ namespace aspect
        */
 
       /**
-       * @name Functions used in the assembly of linear systems
+       * @name Functions, classes, and variables used in the assembly of linear systems
        * @{
        */
+
+      /**
+       * A structure that consists of member variables representing
+       * each one or more functions that need to be called when
+       * assembling right hand side vectors, matrices, or complete linear
+       * systems. We use this approach in order to support the following
+       * cases:
+       * - Assembling different formulations: When assembling either the
+       *   full equations or only the Boussinesq approximation (to give just
+       *   two examples), one needs different terms. This could be achieved
+       *   using a large number of <code>switch</code> or <code>if</code>
+       *   statements in the code, or one could encapsulate each equation
+       *   or approximation into a collection of functions for this particular
+       *   purpose. The approach chosen here in essence allows the
+       *   implementation of each set of equations in its own scope, and we
+       *   then just need to store a pointer to the function that
+       *   assembles the Stokes system (for example) for the selected
+       *   approximation. The pointer to this function is stored in the
+       *   appropriate member variable of this class.
+       * - Sometimes, we want to assemble a number of terms that build on
+       *   each other. An example is the addition of free boundary terms
+       *   to the Stokes matrix. Rather than having to "know" in one
+       *   place about all of the terms that need to be assembled,
+       *   we simply add the function that computes these terms as
+       *   another "slot" to the appropriate "signal" declared in this
+       *   class.
+       */
+      struct Assemblers
+      {
+        boost::signals2::signal<void (const double,
```

Comment:
would it make sense to include the current cell also for the stokes preconditioner?